README.md

```diff
@@ -1,66 +1,145 @@
+* https://github.com/kubernetes-csi/external-snapshotter/tree/master/deploy/kubernetes/snapshot-controller
+* Do this once per cluster
 
-```bash
-for i in $(find deploy/kubernetes -name '*.yaml'); do kubectl create -f $i; done
-```
+Install CSI Driver:
+* Follow instructions provided by your CSI Driver vendor.
+* Here is an example to install the sample hostpath CSI driver
+  * kubectl create -f deploy/kubernetes/csi-snapshotter
+  * https://github.com/kubernetes-csi/external-snapshotter/tree/master/deploy/kubernetes/csi-snapshotter
 
-### Running with Leader Election
+### Important changes in the snapshot beta APIs
+
+* DeletionPolicy is now a required field rather than optional in both VolumeSnapshotClass and VolumeSnapshotContent. This way the user has to explicitly specify it, leaving no room for confusion.
+* VolumeSnapshotSpec has a new required Source field. Source may be either a PersistentVolumeClaimName (if dynamically provisioning a snapshot) or VolumeSnapshotContentName (if pre-provisioning a snapshot).
+* VolumeSnapshotContentSpec also has a new required Source field. This Source may be either a VolumeHandle (if dynamically provisioning a snapshot) or a SnapshotHandle (if pre-provisioning volume snapshots).
+* VolumeSnapshotStatus now contains a BoundVolumeSnapshotContentName to indicate the VolumeSnapshot object is bound to a VolumeSnapshotContent.
+* VolumeSnapshotContentnow contains a Status to indicate the current state of the content. It has a field SnapshotHandle to indicate that the VolumeSnapshotContent represents a snapshot on the storage system.
+
+### Snapshot controller command line options
+
+#### Important optional arguments that are highly recommended to be used
+* `--leader-election`: Enables leader election. This is useful when there are multiple replicas of the same snapshot controller running for the same Kubernetes deployment. Only one of them may be active (=leader). A new leader will be re-elected when current leader dies or becomes unresponsive for ~15 seconds.
+
+* `--leader-election-namespace <namespace>`: The namespace where the leader election resource exists. Defaults to the pod namespace if not set.
+
+#### Other recognized arguments
+* `--kubeconfig <path>`: Path to Kubernetes client configuration that the snapshot controller uses to connect to Kubernetes API server. When omitted, default token provided by Kubernetes will be used. This option is useful only when the snapshot controller does not run as a Kubernetes pod, e.g. for debugging.
+
+* `--resync-period <duration>`: Internal resync interval when the snapshot controller re-evaluates all existing `VolumeSnapshot` instances and tries to fulfill them, i.e. create / delete corresponding snapshots. It does not affect re-tries of failed calls! It should be used only when there is a bug in Kubernetes watch logic.
+
+* `--create-snapshotcontent-retrycount`: Number of retries when we create a snapshot content object for a snapshot. Default is 5.
```

Comment:
does that mean we give up entirely after 5? I thought we changed it to retry forever until user cancels the operation

Reply:
This retry count is used for creating VolumeSnapshotContent API objects.  We didn't talk about retrying this one forever.

The logic we are trying to change is for creating a physical snapshot on the storage system itself.  Grant started a PR which followed what Jan has done for the external-provisioner.  Due to the beta splitting controller work, that work is delayed.  That involves big controller logic changes which I do not want to add on top of the splitting controller changes in this release.  We can revisit that after we cut this release.

Reply:
If API server is temporarily down, then we will give up on trying to bind or create a snapshot? I'm not sure we want that? Also keep in mind especially now that this is beta, we need to support the argument according to the deprecation policy.

Reply:
I think we copied it from the PV controller. Looks like PV controller retries creating PV object 5 times by default, but createProvisionedPVRetryCount is a const defined inside the PV controller.  Should we get rid of these command line options and use a const in the snapshot controller instead?

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L131
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L1489

Reply:
Yeah I think that would be better. It gives more flexibility to change the behavior later.

I think PV controller scenario is slightly different than snapshot. For snapshots, I think we have stronger idempotency/naming guarantees, so it can retry the calls with the same id, which helps (but not completely) avoids the leaking issue. So I'm not sure how necessary the retries in snapshot controller should be. cc @jsafrane

Reply:
Actually, does the scenario apply at all? Now we create the Content object before calling CSI. This avoids the leaking scenario right?

Reply:
Yes, you are right.

Reply:
Removed these from README.  Will submit a PR that removes them from cmd line options and uses a const internally instead.